### ISSUE TITLE
Accessibility for RenderWhiteSpace broken

### ISSUE THREAD

**Issue by krperry** [microsoft/vscode#150971]
Issue Type: <b>Bug</b>

Currently Render white space is either drawing or putting something in the place of spaces that screen readers cannot speak or render to blind and low vision users.

Expected behavior. When RenderWhiteSpace is turned to all the following could be used to replace the space characters so that Screen readers will speak them. This works in other editors by doing this. Note that this is just an example of what could be used. If you think better symbols would work that is fine if both jaws and NVDA speak the characters. The following characters have been evaluated and work.

Space – 0x2024 says <dot>
End of line – 0x2193 says <downwards arrow>
Tab - 0x2192 says <rightwards arrow> or maybe 2023 says <arrow bullet>

It would be nice if the RenderWhiteSpace is pre-defined as a key press to toggle on and off to what ever the setting is set to. This way a blind person could press it and check the indentation and quickly press it to turn off the visible spaces. If the renderWhiteSpace is able to be bound by a user, I guess the default key press does not have to be defined. but it would be nice.

VS Code version: Code 1.67.2 (c3511e6c69bb39013c4a4b7b9566ec1ca73fc4d5, 2022-05-17T18:15:52.058Z)
OS version: Windows_NT x64 10.0.19044
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz (8 x 2904)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.88GB (23.52GB free)|
|Process Argv|--crash-reporter-id ea0256fe-448a-4761-b67e-49aad19149df|
|Screen Reader|yes|
|VM|0%|
</details><details><summary>Extensions (6)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-intelephense-client|bme|1.8.2
python|ms-|2022.6.3
vscode-pylance|ms-|2022.5.3
jupyter|ms-|2022.4.1021342353
jupyter-keymap|ms-|1.0.0
jupyter-renderers|ms-|1.0.8

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyl392:30443607
pythontb:30283811
pythonptprofiler:30281270
vsdfh931:30280409
vshan820:30294714
vstes263:30335439
pythondataviewer:30285071
vscod805:30301674
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
vsc1dst:30438360
pythonvs932:30410667
wslgetstarted:30449410
vscscmwlcmt:30465135
cppdebug:30492333
vsclangdc:30486549

```

</details>

We have written the needed data into your clipboard because it was too large to send. Please paste.

**Comment by isidorn** [comment]
@krperry thanks for opening this issue. Somewhat related to https://github.com/microsoft/vscode/issues/88877

I agree that we should improve here. So you are suggesting that we introduce a command that would read the indentation.
Would you expect this command to just read the indentation, and not the rest of the line correct? And only for the current line?
I am not sure if this should be connected to the renderWhiteSpace setting. I think not.

Is there some other way we could improve the indentation experience in VS Code?
What do Screen Readers currently support out of the box? Can't you say to the screen reader to read the identation?

fyi @jhomme @jvesouza for thoughts

Assigning also to @hediet for the editor side.

**Comment by jvesouza** [comment]
Currently I configure orca, the screen reader I use, to announce the amount of spaces or tabs at the beginning of each line.
The line:
'    x = 1;0'
Is read by orca as:
'four spaces x = 10;'

This configuration can be toggled on or off.

**Comment by krperry** [comment]
I can also do indention with a screen reader .  Your indention settings on orca does not show the spaces in a line for example if you have 4 spaces between 

 

Def myfunc    ():

 

Your screen reader will not tell you that.  You can see it in braille but not if you’re looking with speech and not everyone has a braille display.  You can with this RenderWhiteSpace functionality.  Your screen reader also will not tell you if you have tabs or spaces at the end of the line.  this RenderWhiteSpace does tell you that if it is accessible.    Finally just because you can do it with Orca, Jaws, and NVDA for indention only you can not do it with some of the other screen readers like Narrator which is a free one on windows

**Comment by krperry** [comment]
You do not have to read the white space if you make the characters that show the white space something that the screen readers can read.  This is not just about white space if you know what the RenderWhiteSpace command does it shows you, new lines, tabs, spaces as different at least my current editor does.  For example my screen reader on boxer editor (by the ay this is free to download if you want to try it).  This is what I hear on it when I press alt-f1 and alt-f1 to toggle it on boxer.

 

….def.myfunc.():[

 

If I have tabs instead of spaces as indentation I hear this

 

]def myfunc.():[

 

Note the characters are not normal characters they are just characters that the screen readers can read that is why I put the Unicode characters.  Again this has to do with all white space not just the indention.  If you want another example is look at how reveal codes works in most word processors.  It actually puts codes that screen readers can read.  What ever you are doing now can not be read by a screen reader.

**Comment by isidorn** [comment]
VoiceOver reads now the indentation on mac. Orca supports it already on Linux. And I believe for nvda it would be best to reach out to them. But no work planned on the VS Code side, thus closing
